File: application/src/main/java/org/sheedon/common/handler/SoftKeyboardStateHandler.java
package org.sheedon.common.handler;

import android.graphics.Rect;
import android.view.View;
import android.view.ViewTreeObserver;

import java.util.LinkedList;
import java.util.List;

/**
 * 软键盘监听者
 *
 * @Author: sheedon
 * @Email: sheedonsun@163.com
 * @Date: 2021/8/3 2:35 下午
 */
public class SoftKeyboardStateHandler implements ViewTreeObserver.OnGlobalLayoutListener {
    public interface SoftKeyboardStateListener {
        void onSoftKeyboardOpened(int keyboardHeightInPx);

        void onSoftKeyboardClosed();
    }

    private final List<SoftKeyboardStateListener> listeners = new LinkedList<>();
    private View activityRootView;
    private boolean isSoftKeyboardOpened;

    public SoftKeyboardStateHandler(View activityRootView) {
        this(activityRootView, false);
    }

    public SoftKeyboardStateHandler(View activityRootView, boolean isSoftKeyboardOpened) {
        this.activityRootView = activityRootView;
        this.isSoftKeyboardOpened = isSoftKeyboardOpened;
        activityRootView.getViewTreeObserver().addOnGlobalLayoutListener(this);
    }

    @Override
    public void onGlobalLayout() {
        final Rect r = new Rect();
        //r will be populated with the coordinates of your view that area still visible.
        activityRootView.getWindowVisibleDisplayFrame(r);

        final int heightDiff = activityRootView.getRootView().getHeight() - (r.bottom - r.top);
        if (!isSoftKeyboardOpened && heightDiff > 300) { // if more than 100 pixels, its probably a keyboard...
            isSoftKeyboardOpened = true;
            notifyOnSoftKeyboardOpened(heightDiff);
        } else if (isSoftKeyboardOpened && heightDiff < 300) {
            isSoftKeyboardOpened = false;
            notifyOnSoftKeyboardClosed();
        }
    }

    public void addSoftKeyboardStateListener(SoftKeyboardStateListener listener) {
        listeners.add(listener);
    }

    public void clear() {
        listeners.clear();
        activityRootView = null;
    }

    private void notifyOnSoftKeyboardOpened(int keyboardHeightInPx) {
        for (SoftKeyboardStateListener listener : listeners) {
            if (listener != null) {
                listener.onSoftKeyboardOpened(keyboardHeightInPx);
            }
        }
    }

    private void notifyOnSoftKeyboardClosed() {
        for (SoftKeyboardStateListener listener : listeners) {
            if (listener != null) {
                listener.onSoftKeyboardClosed();
            }
        }
    }

    public boolean isSoftKeyboardOpened() {
        return isSoftKeyboardOpened;
    }
}

File: application/src/main/java/org/sheedon/common/handler/SoftKeyboardStateHandler.kt
package org.sheedon.common.handler

import android.graphics.Rect
import android.view.View
import android.view.ViewTreeObserver
import java.util.*

/**
 * 软键盘监听者
 *
 * @Author: sheedon
 * @Email: sheedonsun@163.com
 * @Date: 2022/1/6 12:07 下午
 */
class SoftKeyboardStateHandler(
    _activityRootView: View,
    _isSoftKeyboardOpened: Boolean = false
) :
    ViewTreeObserver.OnGlobalLayoutListener {

    private val listeners = LinkedList<SoftKeyboardStateListener>()
    private var activityRootView: View? = _activityRootView
    private var isSoftKeyboardOpened = _isSoftKeyboardOpened

    init {
        activityRootView?.viewTreeObserver?.addOnGlobalLayoutListener(this)
    }

    override fun onGlobalLayout() {
        val r = Rect()
        activityRootView?.getWindowVisibleDisplayFrame(r)
        val heightDiff = activityRootView?.rootView?.height?.minus((r.bottom - r.top))
        if (heightDiff != null) {
            if (!isSoftKeyboardOpened && heightDiff > 300) { // if more than 100 pixels, its probably a keyboard...
                isSoftKeyboardOpened = true
                notifyOnSoftKeyboardOpened(heightDiff)
            } else if (isSoftKeyboardOpened && heightDiff < 300) {
                isSoftKeyboardOpened = false
                notifyOnSoftKeyboardClosed()
            }
        }
    }

    fun addSoftKeyboardStateListener(listener: SoftKeyboardStateListener) {
        listeners.add(listener)
    }

    private fun notifyOnSoftKeyboardOpened(keyboardHeightInPx: Int) {
        listeners.forEach {
            it.onSoftKeyboardOpened(keyboardHeightInPx)
        }
    }

    private fun notifyOnSoftKeyboardClosed() {
        listeners.forEach {
            it.onSoftKeyboardClosed()
        }
    }

    fun isSoftKeyboardOpened() = isSoftKeyboardOpened

    fun clear() {
        listeners.clear()
        activityRootView = null
    }


    interface SoftKeyboardStateListener {
        fun onSoftKeyboardOpened(keyboardHeightInPx: Int)

        fun onSoftKeyboardClosed()
    }
}